### PR TITLE
WIP - Fixes #12389 - added image-deployment service

### PR DESCRIPTION
**Pull request by lzap**
Image-based deployment via discovery

_Option A: stream-based_

This adds a service template that starts netcat for each block device
(excluding loops) on particular fixed port. This allows anyone to write to
block devices remotely effectively doing image deployment. The service only
allows writing, not reading, therefore this should be secure. But by default,
the ports are blocked by firewall and `fdi.exposebdev` must be set to `1` in
order to have the devices exposed.

Typical usage:

```
virt-builder fedora-22 --output /tmp/f22.img --size 123450b
```

Where 123450 is full size of the block device (do not forget `b` suffix) as
reported by facter. Image can be smaller (and FS expanded later on) of course.

```
cat /tmp/f22.img | lzop | ncat 192.168.122.68 13200
```

Lzop compressor is a must, the endpoint expect stream to be compressed with lzop.
Port 13200 is the port for `/dev/vda` since in my case this was a VM created with KVM. 
See the mapping table below.

_Option B: UDP cast - unicast or multicast_

This uses https://en.wikipedia.org/wiki/UDPCast utility that is in EPEL (actually the package is pending in testing repo: https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2015-ef035e942f). Read the project page how it works or watch our Community demo from 10.12.2015.

Usage is simple, enable this service by providing device name without `/dev/` prefix with this kernel option: `fdi.udpcast=vda`. The boot as many discovered hosts as you want and once you are ready to send the image do:

```
udp-sender /tmp/cirros.img --interface virbr0
```

I need to specify the device as I use libvirt virtual network. Make sure your network allows multicast UDP packets. When using just one host, udpcast uses unicast.

The port mapping for option A is:

| dev | port |
| --- | --- |
| sda | 13000 |
| sdb | 13001 |
| sdc | 13002 |
| sdd | 13003 |
| sde | 13004 |
| sdf | 13005 |
| sdg | 13006 |
| sdh | 13007 |
| sdi | 13008 |
| sdj | 13009 |
| sdk | 13010 |
| sdl | 13011 |
| sdm | 13012 |
| sdn | 13013 |
| sdo | 13014 |
| sdp | 13015 |
| sdq | 13016 |
| sdr | 13017 |
| sds | 13018 |
| sdt | 13019 |
| sdu | 13020 |
| sdv | 13021 |
| sdw | 13022 |
| sdx | 13023 |
| sdy | 13024 |
| sdz | 13025 |
| hda | 13100 |
| hdb | 13101 |
| hdc | 13102 |
| hdd | 13103 |
| hde | 13104 |
| hdf | 13105 |
| hdg | 13106 |
| hdh | 13107 |
| hdi | 13108 |
| hdj | 13109 |
| hdk | 13110 |
| hdl | 13111 |
| hdm | 13112 |
| hdn | 13113 |
| hdo | 13114 |
| hdp | 13115 |
| hdq | 13116 |
| hdr | 13117 |
| hds | 13118 |
| hdt | 13119 |
| hdu | 13120 |
| hdv | 13121 |
| hdw | 13122 |
| hdx | 13123 |
| hdy | 13124 |
| hdz | 13125 |
| vda | 13200 |
| vdb | 13201 |
| vdc | 13202 |
| vdd | 13203 |
| vde | 13204 |
| vdf | 13205 |
| vdg | 13206 |
| vdh | 13207 |
| vdi | 13208 |
| vdj | 13209 |
| vdk | 13210 |
| vdl | 13211 |
| vdm | 13212 |
| vdn | 13213 |
| vdo | 13214 |
| vdp | 13215 |
| vdq | 13216 |
| vdr | 13217 |
| vds | 13218 |
| vdt | 13219 |
| vdu | 13220 |
| vdv | 13221 |
| vdw | 13222 |
| vdx | 13223 |
| vdy | 13224 |
| vdz | 13225 |

### CODE REVIEW

**Comment by lzap** [comment]
This one is WIP, for discussion. I may want to replace gzip with faster lzop and also base this off http://www.udpcast.linux.lu/cmdlinedoc.html (but that's not in CentOS7/RHEL7). Could be an alternative.

**Comment by lzap** [comment]
Looks like `lzop` is SO much faster, anyway I have idea on a different approach: nbd with virt-builder. DO NOT MERGE. WIP.

File: 00-repos-centos7.ks [hunk 1]
```diff
@@ -3,3 +3,4 @@ repo --name=centos-updates --mirrorlist=http://mirrorlist.centos.org/?release=7&
 repo --name=epel7 --mirrorlist=https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=$basearch
 repo --name=foreman-el7 --baseurl=http://yum.theforeman.org/nightly/el7/$basearch/
 repo --name=foreman-plugins-el7 --baseurl=http://yum.theforeman.org/plugins/nightly/el7/$basearch/
+repo --name=wip-epel7-udpcast --baseurl=https://copr-be.cloud.fedoraproject.org/results/lzap/udp-cast-epel7/epel-7-$basearch/
```
Comment (lzap) [inline 00-repos-centos7.ks]: This is temporary solution until udpcast gets it into EPEL7 (currently in testing). Do not merge, but please review.

**Comment by lzap** [comment]
Open ideas:

* TUI for this
* support in Discovery Plugin
* ?

**Comment by beddari** [comment]
As a side note, I think an idea to consider would be to expand this to do "machine zero in the DC" "self-deployments", e.g where you boot the iso using ILO/iDRAC type APIs and install Foreman to the same host.

**Comment by lzap** [comment]
Can you elaborate this please?

File: 00-repos-centos7.ks (hunk 1 above)
Comment (lzap) [inline 00-repos-centos7.ks]: https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2015-ef035e942f

The package is now in EPEL6/7...

**Comment by lzap** [comment]
For the record, I submitted a talk about this at CfgMgmt camp in Berlin 2016. The patch works fine, one just needs to issue image transfer manually. The idea is to build UI in Foreman for this - kind of an image store (initial implementation just raw images in a directory) and virt-builder helper utilities to build initial set of images with finish script firstboot call. Support in the image would include partitioning disks and fetching images either directly or via udpcast for multiple servers.

**Comment by beddari** [comment]
That is nice Lukas! I'm still interested in this use case, hope to spend more time on it ... (and then yes a talk in Ghent would help!)

**Comment by Raboo** [comment]
Just wanted to pitch in some ideas.
Why the whole netcat complication?

Have a cloud-init aware image and run curl or wget?

`curl http://somehost/image.img | dd bs=2M of=/dev/sda`
`wget -O - http://somehost/image.img | dd bs=2M of=/dev/sda`

**Comment by lzap** [comment]
New version is referenced from this PR, netcat removed, udpcast and HTTP transfer still an option.

**Pull request closed without merging**